torch/utils/_sympy/reference.py

```diff
@@ -70,7 +70,10 @@ def square(x):
 
     @staticmethod
     def mod(x, y):
-        return x % y
+        ret = abs(x) % abs(y)
+        if x < 0:
```

Comment:
What's the problem with python's definition of mod? Also isn't this supposed to work with sympy values, so we can't do an if?

Reply:
This PR implements mod as C's mod, as it's the one that's used in codegen. Now, if we are also using Python expressions for shape reasoning, then we actually need to implement both.

At the moment, our split between "which sympy classes represent Python syntax" and "which sympy classes represent C syntax" is non-existent. Perhaps a better fix would be to actually try to do this split properly in the first place.

Reply:
> Also isn't this supposed to work with sympy values, so we can't do an if?

It'd be good if it did, but truncation doesn't work yet either (here we are doing something derived from truncation really) because SymPy does not have a truncation op that supports symbols. We could implement one, but we don't have one.

Reply:
AFAICT they are the same?
```python
In [1]: import torch
   ...: a = torch.arange(-10, 10)
   ...: 
   ...: torch.testing.assert_close(a % 3, torch.tensor([x % 3 for x in range(-10, 10)]))
   ...: torch.testing.assert_close(a % -3, torch.tensor([x % -3 for x in range(-10, 10)]))

```

Reply:
Yes, PyTorch mimics Python. What I mean here is that `Mod` / `ModularIndexing` and the like we are lowering them to `%` in both Python and C++. So there are a number of possibilities here:
1. Our IR maintains Python semantics
2. Our IR maintains C semantics
3. We have different IRs for each

Of the three, 1 is the worst one, as we are going to generate slow code whenever we or LLVM can't optimise Python mod -> C mode. It also goes against the whole IR, where we represent objects in a way that's amenable to codegen.

In general, I think that option 3 is the best here. Guards and wrapper coe will need to use Python mod and codegen will need to use C mod.

This is going to be awful to get right, but I think it's something we should do, as otherwise it would lead to some rather obscure errors.

Thoughts on priority @ezyang?

p.s. if we deem that this is not a priority, I can implement regular Python value ranges in this PR and move on.

Reply:
Should have both (remainder and fmod). I guess one can quibble about what mod should default to. I think we have a precedent with div though: goes to truediv (Python) not truncdic (C)

Reply:
>I think we have a precedent with div though: goes to truediv (Python) not truncdic (C)

That's just partially true. We do treat them as python semantics during all the analysis, but then at codegen we do
```python
    def _print_FloorDiv(self, expr):
        if expr.is_integer:
            return super()._print_FloorDiv(expr)
```
where the parent is the Python printer... ooops!

same same for ModularIndexing:
```python
    def _print_ModularIndexing(self, expr):
        x, div, mod = expr.args
        x = self.paren(self.doprint(x))
        div = self.paren(self.doprint(div))
        mod = self.paren(self.doprint(mod))
        if div != "1":
            x = f"({x} // {div})"
        return f"{x} % {mod}"
```
Worst of all, we cannot fix this as the perf hit would be quite serious. At the moment this sort of works because in most cases these ops are applied to positive inputs, but yeah...

Reply:
I see, I forgot that `__mod__` is actually `remainder` and not `fmod` so was looking at the wrong eager kernel. Looks like the current state is `ops.mod` -> C Modulo, `ops.remainder` -> Python Modulo which makes sense given that `fmod`, `remainder` names in aten so I think this is fine.

> I think we have a precedent with div though: goes to truediv (Python) not truncdic (C)

tbh I think it's a bit silly that `ops.div` exists at all. imo we should just explicitly spell out `truediv` and `truncdiv` rather than having ops that are meant to be equivalent.

Reply:
The small issue is that we have:
- PyTorch frontend
- ops.*
  - lowered to C++ and triton
- sympy classes
  - lowered to Python (wrapper / guards) AND C++ & triton (C semantics)

PyTorch -> ops is reasonable consistent so that's nice. Now, the SymPy story is all over the place. That's the one we should get right.

Regarding op.div, I completely agree. IMO, we should have three: 
- `ops.truediv` equiv to `__div__`
- `ops.floordiv` equiv to `__floordiv__`
- `ops.truncdiv` equiv to `/` in C

Reply:
Well, I am happy to not have the mod name. But we still have a problem on how to compile modular indexing. 

The Correct(TM) fix here I think involves value range analysis. In particular, when lhs/rhs is known to be positive, we can elide the abs. For sizey things, specifically indexing compute, we typically know this and the elision should work. I don't have too many opinions about how exactly we should so it, but it seems feasible. WDYT?

Reply:
We already do that to some extent here:
https://github.com/pytorch/pytorch/blob/512759a3d7aa064abff6b7e7be720f84d5f5764e/torch/_inductor/index_propagation.py#L126-L141

A start would be to have a Mod and a Remainder class in SymPy. Also declare that ModularIndexing has C semantics for the division and the mod, given that is an Inductor class. Then, in the wrapping printing we can do a value range analysis and see if we can just write (a // b) % c in Python, and if we don't know, print the proper expressions.

next two weeks I'll be on PTO, but I can do that when I'm back